### PR TITLE
Issue 1215:  Segment Truncation (Part 1)

### LINKED ISSUE #1215: Segment Truncation
**Problem description**
We need to support compaction and more fine grained retention policies. As part of this, we should support truncation of all data before a supplied offset on a per segment basis. See #284 for more details.

**Problem location**
StreamSegmentStore

**Suggestions for an improvement**
Provide an API in `StreamSegmentStore` along the lines of:
`CompletableFuture<Void> truncateStreamSegment(String streamSegmentName, long offset);`

Details:
- This cannot be applied to transactions.
- Upon completion, this will set an internal attribute (or some other bit) on the Segment which indicates the lowest available offset for reading.
- For correctness, a Segment can only be truncated progressively (i.e., once we truncate at offset A, we can only truncate at offset B>=A).
- Once a Segment is truncated in its entirety (truncation offset >= Segment length), it should be deleted.
- Reads below the truncation offset are not allowed.
    - In-progress reads for a Segment during truncation will be canceled if their current ReadOffset is below the Truncation offset. 
    - Otherwise, the reads will proceed normally.
    - Since technically every offset in the segment is still accessible (in Tier2), it should be sufficient to do the offset check in the `StreamSegmentReadIndex` class.
- In order to maintain our consistency guarantees, we should not do any offset adjustment. Once set, the value at a particular offset should never change. As such, upon truncation, the truncation offset will not become offset 0 in the segment. Instead, the `GetSegmentInfo` method will return an additional field that contains the first offset available for reading. Any access below this offset will not be allowed.

### PR DESCRIPTION
**Pull request by andreipaduroiu**
**Change log description**
Added the ability to truncate a Segment in the Segment Store.

Highlights:
- A segment can be truncated at a particular offset. The initial offset is 0, and it can only be increased. The truncation offset can never exceed the current length of the Segment.
- The Segment Store will disallow access to any bytes before the truncation offset. 
- Append behavior will not change. Appends only happen at the end of the Segment, and by definition, truncations cannot interfere with this.
- Read behavior
    - New reads will return a `ReadResult`, but the first `ReadResultEntry` in that `ReadResult` is of type `Truncated`. 
    - Ongoing reads will return a `Truncated` `ReadResultEntryType` if the current read offset has been truncated. Any attempt to move on or get contents of this entry will fail with `StreamSegmentTruncatedException`. 
- `GetStreamSegmentInfo` now contains `getStartOffset` which indicates the current Truncation offset of the segment.
- How to use this
    - `PravegaRequestProcessor` can issue a `GetStreamSegmentInfo` prior to any reads to get the Truncation offset. It can also do the same when it encounters a `ReadResultEntryType.Truncated` read entry.
- Restrictions: 
    - Transactions cannot be truncated
    - There is no restriction as to whether a Segment needs to be sealed or not prior to truncation.
    - This does not affect the state of the Segment in Tier2. As of now, there is no Tier2 implementation that natively supports truncation at the beginning. If desired, this will be tackled in a future issue.

Code Changes:
- Added `StreamSegmentStore.truncateStreamSegment(SegmentName, Offset)` as a public API
- Added `StreamSegmentTruncateOperation` which implements the above API. This is processed in the same manner as all other operations in the SegmentStore and provides the same guarantees.
- Updated `SegmentProperties` to include `getStartOffset` which, after the successful execution of a Truncate Operation, will contain the new truncation offset.
- Updated Read Index to handle the truncation behavior.
- Storing Truncation Offset in the `SegmentState` when the segment becomes inactive and its metadata is evicted, and reloading it from there.

**Purpose of the change**
Fixes #1215.

**What the code does**
All details in **Change Description**

**How to verify it**
Lots of new unit tests added.

### CODE REVIEW

**Comment by tkaitchuck** [comment]
Rather than using IllegalArgumentException can we create a specific exception. (It can still be runtime)

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java [hunk 1]
```diff
@@ -85,8 +89,9 @@ public static SegmentState deserialize(DataInputStream source) throws IOExceptio
             if (version == SERIALIZATION_VERSION) {
                 long segmentId = source.readLong();
                 String segmentName = source.readUTF();
+                long startOffset = source.readLong();
                 Map<UUID, Long> attributes = AttributeSerializer.deserialize(source);
-                return new SegmentState(segmentId, segmentName, attributes);
+                return new SegmentState(segmentId, segmentName, startOffset, attributes);
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java]: Does this need to bump the Serializetion_version above?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java [hunk 2]
```diff
@@ -370,6 +372,27 @@ public int getId() {
...
+        logRequest("truncateStreamSegment", streamSegmentName);
+        this.metrics.truncate();
+        TimeoutTimer timer = new TimeoutTimer(timeout);
+        AtomicLong segmentId = new AtomicLong();
+        return this.segmentMapper
+                .getOrAssignStreamSegmentId(streamSegmentName, timer.getRemaining(),
+                        streamSegmentId -> {
+                            segmentId.set(streamSegmentId);
+                            StreamSegmentTruncateOperation op = new StreamSegmentTruncateOperation(streamSegmentId, offset);
+                            return this.durableLog.add(op, timer.getRemaining());
```
Comment (tkaitchuck) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: Here this is returning a CompletableFuture<Void> from another method but setting segmentId to pass it to the thenApply block below. Ideally we would chain a future that returned segmentId on the result of durableLog.add

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java (hunk 1 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java]: For this version we don't support backward compatibility as far as I know

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java (hunk 2 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: `durableLog.add()` returns a `CompletableFuture<Void>` which completes when the operation has been processed. There are 10 types of operations, most of which do not deal with Segment Ids. As such, I can't have it return a SegmentId, so I need to extract it from the operation after its completion.

**Comment by andreipaduroiu** [comment]
@tkaitchuck 
> Rather than using IllegalArgumentException can we create a specific exception. (It can still be runtime)

It uses `IllegalArgumentException` when you pass a wrong argument to `read`, in order to be consistent with the other checks it does (since in the end, it's an invalid argument). There is a `StreamSegmentTruncatedException` that is thrown for all in-progress reads that are truncated out. In addition, a new type of `ReadResultEntryType` was added so that the upstream layers can figure out when their reads are interrupted and need to take action.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java (hunk 2 above)
Comment (tkaitchuck) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/StreamSegmentContainer.java]: I was not suggesting to change add() but rather to take it's result and compose it with a completed future of the segmentId and return that.

**Comment by tkaitchuck** [comment]
I'm not sure it is (or should be) illegal for readers to attempt to read from below the truncated point. They simpliy need to be notified, and IllegalArgument is very difficult to catch and handle because any number of other things could have caused it. 
Consider a reader reads 1-100 on a segment that goes up to 1000 then someone comes and truncates at 500. The reader had no way of knowing that that occurred. Even if it checked before every single read there is still a race. So when it tries to read 100-200 we need to tell it what happened, not simply log IllegalArgument and close its connection.

**Comment by andreipaduroiu** [comment]
So the way `ReadResults` work is that they behave like an iterator. Every time you call `next` you get some `ReadResultEntry` back. This entry can be of multiple types (`Cache`, `Storage`, `Future`, `EndOfSegment`) and now I've added `Truncated`. The WireProtocol server-side code can check this (It already does, for `EndOfSegment`) and if it is `Truncated`, then it can decide what to do. At that point, since the `TruncateOperation` has already been processed, the data may already be gone from Tier1, the cache and possibly from Tier2. I don't know what else I can return, since I can't block the `TruncateOperation` on every possible reader to finish what they're doing.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java (hunk 1 above)
Comment (tkaitchuck) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java]: Isn't the version used to check for backwards compatibility? So shouldn't it be incremented precisely because this change is breaking backward compatibility?

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentStore.java [hunk 3]
```diff
@@ -161,4 +161,18 @@
      * @throws IllegalArgumentException If any of the arguments are invalid.
      */
     CompletableFuture<Void> deleteStreamSegment(String streamSegmentName, Duration timeout);
+
+    /**
+     * Truncates a Sealed StreamSegment at a given offset.
```
Comment (fpj) [inline segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentStore.java]: I'm confused about this comment, the PR description says:

```
There is no restriction as to whether a Segment needs to be sealed or not prior to truncation.
```

What am I missing?

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentNotSealedException.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+     *
+     */
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Creates a new instance of the StreamSegmentNotSealedException class.
+     *
+     * @param streamSegmentName The name of the StreamSegment that is not sealed.
+     */
+    public StreamSegmentNotSealedException(String streamSegmentName) {
```
Comment (fpj) [inline segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentNotSealedException.java]: Is this only used with txns?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java (hunk 1 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java]: No, I meant that for our version of Pravega we don't care about backwards compatibility. 

Especially when Part 2 of this comes in and it will require Tier2 to be wiped out and started clean because of the new file layout.

**Comment by andreipaduroiu** [comment]
The changes to `PravegaRequestProcessor` and upstream (Client) can be addressed as a separate issue. This PR only addresses the functionality in the SegmentStore

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/operations/StreamSegmentTruncateOperationTests.java [hunk 5]
```diff
@@ -0,0 +1,24 @@
+/**
+ * Copyright (c) 2017 Dell Inc., or its subsidiaries. All Rights Reserved.
+ * <p>
```
Comment (fpj) [inline segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/operations/StreamSegmentTruncateOperationTests.java]: Please revert the changes to license header.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java [hunk 6]
```diff
@@ -318,31 +317,29 @@ void preProcessOperation(Operation operation) throws ContainerException, StreamS
...
-            } else if (operation instanceof UpdateAttributesOperation) {
-                segmentMetadata.preProcessOperation((UpdateAttributesOperation) operation);
-            } else if (operation instanceof MetadataCheckpointOperation) {
-                // MetadataCheckpointOperations do not require preProcess and accept; they can be handled in a single stage.
-                processMetadataOperation((MetadataCheckpointOperation) operation);
-            } else if (operation instanceof StorageMetadataCheckpointOperation) {
-                // StorageMetadataCheckpointOperation do not require preProcess and accept; they can be handled in a single stage.
-                processMetadataOperation((StorageMetadataCheckpointOperation) operation);
-            }
+        if (operation instanceof StreamSegmentAppendOperation) {
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java]: For my understanding only: why are we not making a distinction between `StorageOperation` and `MetadataOperation` any longer?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java [hunk 7]
```diff
@@ -481,6 +476,12 @@ private void updateMetadata(StreamSegmentMapping mapping, UpdateableSegmentMetad
 
         // Length must be at least StorageLength.
         metadata.setLength(Math.max(mapping.getLength(), metadata.getLength()));
+
+        // StartOffset must not exceed the last offset of the Segment.
+        if (metadata.getLength() > 0) {
+            metadata.setStartOffset(Math.min(mapping.getStartOffset(), metadata.getLength() - 1));
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java]: The min call here implies that it is possible that `mapping.getStartOffset() > metadata.getLength() - 1`. How can we get to such a scenario?

File: shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java [hunk 8]
```diff
@@ -48,6 +48,7 @@
     public static final String CONTAINER_CREATE_TXN_COUNT = "container_create_txn_count";
     public static final String CONTAINER_MERGE_TXN_COUNT = "container_merge_txn_count";
     public static final String CONTAINER_SEAL_COUNT = "container_seal_count";
+    public static final String CONTAINER_TRUNCATE_COUNT = "container_truncate_count";
```
Comment (fpj) [inline shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: Is this the number of truncate operations?

Comment (andreipaduroiu) [reply shared/metrics/src/main/java/io/pravega/shared/MetricsNames.java]: yes, similar to the ones above it.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java (hunk 6 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java]: it was redundant. It was checking the type of the operation twice, and the first one was redundant. It wasn't doing anything special, so I simplified the code

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentNotSealedException.java (hunk 4 above)
Comment (andreipaduroiu) [reply segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentNotSealedException.java]: It's used whenever a Segment needs to be sealed but it's not. In our case, it's when a merge/concat source is not sealed, so yes, currently only used with transactions.

**Comment by andreipaduroiu** [comment]
@fpj The PR description explains how the Read behaves. I hope this clarification explains how it should work: https://github.com/pravega/pravega/pull/1926#issuecomment-334902113

IMO, It's up to the Client/WireProtocol to figure out how to make use of this information, and that is not addressed in this PR. The SegmentStore provides every information to the upstream layers (via the special ReadResultEntry and via the extra info from GetSegmentInfo) so that the Client/WP can decide what to do next.

**Comment by fpj** [comment]
@andreipaduroiu 
> The PR description explains how the Read behaves.

The description of the PR request does say that it returns `IllegalArgumentException`, but it does not clarify if there can be any ambiguity with respect to this exception. Going through the comments, I believe @tkaitchuck makes a similar point (with a concrete suggestion):

https://github.com/pravega/pravega/pull/1926#issuecomment-333270704

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/ReadResultEntryType.java [hunk 9]
```diff
@@ -9,30 +9,45 @@
...
-    EndOfStreamSegment
+    EndOfStreamSegment(true),
+
+    /**
+     * The ReadResultEntry points to a location in the StreamSegment that has been truncated out and is no longer available.
+     */
+    Truncated(true);
+
+    /**
+     * If true, this indicates that the ReadResult cannot progress from this point anymore and should be closed.
```
Comment (fpj) [inline segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/ReadResultEntryType.java]: I'm not sure I get this comment. It says that if terminal is `true`, then `ReadResult` cannot progress. I don't understand what it means for `ReadResult` to progress. Is it referring to the reader that issue the read making progress?

I searched for `getTerminal` but couldn't find any instance of this getter.

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentNotSealedException.java (hunk 4 above)
Comment (fpj) [reply segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentNotSealedException.java]: ok

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java [hunk 10]
```diff
@@ -200,9 +200,12 @@ public long updateGenerations(int currentGeneration, int oldestGeneration) {
                 // We can only evict if both these conditions are met:
                 // 1. The entry is a Cache Entry (Redirect entries cannot be removed).
                 // 2. Every single byte in the entry has to exist in Storage.
+                // In addition, we are free to evict (regardless of Generation, but still subject to the above rules) if
+                // every single byte in the entry has been truncated out.
+                long lastOffset = entry.getLastStreamSegmentOffset();
                 boolean canRemove = entry.isDataEntry()
-                        && entry.getLastStreamSegmentOffset() <= this.metadata.getStorageLength()
-                        && entry.getGeneration() < oldestGeneration;
+                        && lastOffset <= this.metadata.getStorageLength()
```
Comment (fpj) [inline segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java]: Should it be instead `lastOffset < this.metadata.getStorageLength()`?

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java (hunk 1 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/SegmentState.java]: Fixed. Also filed #1956 to come up with a more elegant solution to this, since we'll need to use this pattern in a lot of places.

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentStore.java (hunk 3 above)
Comment (andreipaduroiu) [reply segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/StreamSegmentStore.java]: Fixed the entire JavaDoc. It was very out of date

File: segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/operations/StreamSegmentTruncateOperationTests.java (hunk 5 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/test/java/io/pravega/segmentstore/server/logs/operations/StreamSegmentTruncateOperationTests.java]: fixed

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java (hunk 7 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java]: When the Segment Is empty (length ==0 ) and we truncate at offset 0.

File: segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/ReadResultEntryType.java (hunk 9 above)
Comment (andreipaduroiu) [reply segmentstore/contracts/src/main/java/io/pravega/segmentstore/contracts/ReadResultEntryType.java]: If your ReadResult, which is an iterator, encounters a `Truncated` ReadResultEntryType, then the ReadResult cannot iterate anymore,therefore it's a terminal entry for that ReadResult

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java (hunk 10 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/StreamSegmentReadIndex.java]: lastOffset == StartOffset + Length, which means `<=` is appropriate. It means the length of the Segment up to the end of that Entry.

**Comment by andreipaduroiu** [comment]
>One additional question I have that I couldn't determine by inspecting the code is the following. Is it possible to have a race, e.g., between a call to GetStreamInfo and a read such the read returns an event in earlier offset compared to the start offset in the segment info (or the other way around)?

GetStreamSegmentInfo returns the current information in the Metadata. As soon as the metadata is updated, any call to `ReadResult.next()` that would serve an offset below `StartOffset` would return a `Truncated ReadResultEntry`. I don't know what you mean by race, because both of these are read operations and don't modify anything. If you call them concurrently, and one happens to serve data prior to the Metadata update and another after the Metadata update, then yes, you may get different results. I can't fix that.

**Comment by andreipaduroiu** [comment]
>The description of the PR request does say that it returns IllegalArgumentException, but it does not clarify if there can be any ambiguity with respect to this exception. Going through the comments, I believe @tkaitchuck makes a similar point (with a concrete suggestion):

This is what the PR says:
>-    New reads will fail with IllegalArgumentException if their read offset is below the truncation offset
>-    Ongoing reads will fail with StreamSegmentTruncatedException if their current read offset is before the truncation offset (a new ReadResultEntryType has been added in order to detect it without having to catch the exception). Otherwise they will continue as usual.

What  that means is that if you issue a read for a truncated offset, you'll get an `IllegalArgumentException`. This is consistent with issuing a read for a negative offset, negative read length, or any other invalid argument. 

If you have an ongoing read, you can inspect the ReadResultEntry you get to figure out what to do 
(there is no exception being thrown at you except if you ignore the entry type and try to request the contents). This is similarly to how the SegmentStore informs the reader about reaching the End-Of-Segment (if sealed) or a Future Read. If you encounter an End-Of-Segment, you can choose to stop there or you can choose to get its contents, in which case you get an exception thrown back at you. For a Future, if you request its contents, you'll get a CompletableFuture back which will complete whenever data is available. As you can see, Truncated data is treated in the same way. The Client/WP has all the info it needs to determine what to do next.

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java (hunk 7 above)
Comment (fpj) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java]: But if `length == 0`, then the condition to execute the if block is false and we will skip it, no?

**Comment by fpj** [comment]
@andreipaduroiu I'm confused about the locations in the code you are referring to, specifically:

> What that means is that if you issue a read for a truncated offset, you'll get an IllegalArgumentException. This is consistent with issuing a read for a negative offset, negative read length, or any other invalid argument.

When you say `you issue a read`,  are you referring to an `EventStreamReader` or `PravegaRequestProcessor`. I think you refer to the latter, but I want to make sure I understand where exactly I should expect the `IllegalArgumentException`.

> If you have an ongoing read, you can inspect the ReadResultEntry you get to figure out what to do
(there is no exception being thrown at you except if you ignore the entry type and try to request the contents).

I suppose that this is saying that I need to inspect `ReadResultEntry` in `PravegaRequestProcessor.handleReadResult, yes?

Also, to clarify, when you say ongoing reads, do you mean reads executed concurrently with the truncation?

**Comment by andreipaduroiu** [comment]
`IllegalArgumentException` is thrown at you when you call `StreamSegmentStore.read` with bad arguments, which includes an offset below the current truncation point. It is not thrown at any other time.

**Comment by tkaitchuck** [comment]
PravegaRequestProcessor calls `read` when the client needs to read data, it has no way of knowing that a segment has been truncated as it not the one performing the truncation. If `ReadResult.next()` threw an exception I would understand. PravegaRequestProcessor is going to need some way of finding out that the read cannot occur because it was truncated.

**Comment by fpj** [comment]
@tkaitchuck I think @andreipaduroiu's argument will be that the segment only knows that the argument is invalid, that such an offset is not currently present in the segment.

Partly, @andreipaduroiu is right that is safe to not make assumptions about why the offset is invalid. At the same time, it sounds safe to assume that a positive offset that is lower than the base offset is not present because it has been truncated, so I can see either the segment store making that assumption or pravega request processor. Having the segment store throwing a truncated exception would reduce the number of steps to figure this out.

**Comment by andreipaduroiu** [comment]
I updated the code to not throw `IllegalArgumentException` anymore; instead it unifies the paths via the `ReadResultEntryType` being set to `Truncated` for the first entry (see description for more details).

File: segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java (hunk 7 above)
Comment (andreipaduroiu) [reply segmentstore/server/src/main/java/io/pravega/segmentstore/server/logs/ContainerMetadataUpdateTransaction.java]: You're right. Then I guess my reasoning for putting it there was to have a failsafe in case upstream code did not do its job correctly. StartOffset should never exceed the Segment's length.
